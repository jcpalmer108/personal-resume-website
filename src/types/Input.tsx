export type InputProps = {
  label: String,
  value?: any,
  updateForm: Function,
  area?: boolean,
  noBorder?: boolean,
}

export type NoBorderProps = {
  noBorder?: boolean 
}

export type LabelProps = {
  paddedLabel?: boolean 
}
