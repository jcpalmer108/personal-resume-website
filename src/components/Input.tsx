import { ChangeEvent } from "react"
import {
  TextField,
  TextArea,
  Label,
  Wrapper
} from "../styles/components/Input"
import { InputProps } from "../types/Input"

export default function Input({ label, value, updateForm, area, noBorder }: InputProps) {
  return (
    <Wrapper data-testid="Input">
      {
        area ? 
          (<TextArea noBorder={noBorder} placeholder=" " value={value} onChange={(e: ChangeEvent) => updateForm(e)} />) : 
          (<TextField noBorder={noBorder} placeholder=" " value={value} onChange={(e: ChangeEvent) => updateForm(e)} />)
      }
      <Label>{label.toUpperCase()}</Label>
    </Wrapper>
  )
}