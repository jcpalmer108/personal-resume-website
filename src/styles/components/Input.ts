import styled from "styled-components";
import {
  silver_polish,
  fortress_gray,
  tabletBreakpoint,
} from "../../styles/variables";
import { LabelProps } from "../../types/Input";

export const Wrapper = styled.div`
  display: flex;
  position: relative;
  flex-direction: column;
  padding: 10px 0px;
`;

export const TextArea = styled.textarea`
  padding: 15px 15px;
  border-radius: 0px;
  border: 1px solid ${silver_polish};
  z-index: 2;
  background: transparent;
  font-family: Gilroy-Regular;
  height: 110px;
  resize: none;
  width: calc(100% - 30px);

  @media (min-width: ${tabletBreakpoint}px) {
    border: 0px;
  }
`;

export const TextField = styled.input`
  padding: 0px 15px;
  height: 50px;
  border-radius: 0px;
  border: 1px solid ${silver_polish};
  z-index: 2;
  background: transparent;
  font-family: Gilroy-Regular;

  @media (min-width: ${tabletBreakpoint}px) {
    border: 0px;
  }
`;

export const Label = styled.label<LabelProps>`
  font-family: Gilroy-Medium;
  font-size: 12px;
  color: ${fortress_gray};
  position: absolute;
  pointer-events: none;
  left: 15px;
  top: 30px;
  transition: 0.2s ease all;
  background: white;
  z-index: 2;

  ${TextField}:not(:placeholder-shown) ~ & {
    top: 3px;
    padding: 3px 15px;
    border-radius: 25px;
    font-size: 10px;
  }

  ${TextField}:focus ~ & {
    top: 3px;
    padding: 3px 15px;
    border-radius: 25px;
    font-size: 10px;
  }

  ${TextArea}:not(:placeholder-shown) ~ & {
    top: 3px;
    padding: 3px 15px;
    border-radius: 25px;
    font-size: 10px;
  }

  ${TextArea}:focus ~ & {
    top: 3px;
    padding: 3px 15px;
    border-radius: 25px;
    font-size: 10px;
  }

  @media (min-width: ${tabletBreakpoint}px) {
    ${TextField}:not(:placeholder-shown) ~ & {
      padding: 3px 10px;
      left: 5px;
    }

    ${TextField}:focus ~ & {
      padding: 3px 10px;
      left: 5px;
    }

    ${TextArea}:not(:placeholder-shown) ~ & {
      padding: 3px 10px;
      left: 5px;
    }

    ${TextArea}:focus ~ & {
      padding: 3px 10px;
      left: 5px;
    }
  }
`;
