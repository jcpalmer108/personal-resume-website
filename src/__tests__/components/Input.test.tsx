import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Input from "../../components/Input";

// params for component
const params = {
  required: {
    label: 'test-label',
    updateForm: jest.fn(),
  },
  optional: {
    value: 'test-value',
    area: true,
    noBorder: true,
  }
}

describe('Input', () => {
  test('renders if optional and required params are passed in', () => {
    // given
    const { required, optional } = params;
    render(<Input label={required.label} value={optional.value} updateForm={required.updateForm} area={optional.area} noBorder={optional.noBorder}/>)
    expect(screen.getByTestId("Input")).toMatchSnapshot()
    expect(screen.getByTestId("Input").childNodes[0].nodeName).toBe('TEXTAREA')
    expect(screen.getByTestId("Input").lastChild.nodeName).toBe('LABEL')
    expect(screen.getByTestId("Input").lastChild).toHaveTextContent(params.label.toUpperCase())
  })

  test('renders if only required params are passed in', () => {
    render(<Input label={params.label} value={params.value} updateForm={params.updateForm}/>)
    expect(screen.getByTestId("Input")).toMatchSnapshot()
    expect(screen.getByTestId("Input").firstChild.nodeName).toBe('INPUT')
    expect(screen.getByTestId("Input").lastChild.nodeName).toBe('LABEL')
    expect(screen.getByTestId("Input").lastChild).toHaveTextContent(params.label.toUpperCase())
  })

  test('updateForm should be called on input update', () => {
    render(<Input label={params.label} value={params.value} updateForm={params.updateForm}/>)
    const inputField = screen.getByTestId("Input")
    userEvent.type(inputField.children[0], "test");
    expect(params.updateForm).toHaveBeenCalledTimes(4); // one call for each letter input
  })

})
