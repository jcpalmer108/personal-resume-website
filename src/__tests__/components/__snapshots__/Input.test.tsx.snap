// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders if optional and required params are passed in 1`] = `
<div
  class="sc-bczRLJ hOHvzZ"
  data-testid="Input"
>
  <textarea
    class="sc-gsnTZi kdzWPC"
    placeholder=" "
  >
    test-value
  </textarea>
  <label
    class="sc-hKMtZM cWKlol"
  >
    TEST-LABEL
  </label>
</div>
`;

exports[`Input renders if required params and value are passed in 1`] = `
<div
  class="sc-bczRLJ hOHvzZ"
  data-testid="Input"
>
  <input
    class="sc-dkzDqf Wkyrw"
    placeholder=" "
    value="test-value"
  />
  <label
    class="sc-hKMtZM cWKlol"
  >
    TEST-LABEL
  </label>
</div>
`;
